Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,329 @@
+  }
+
+  /// A metadata key-value pair.
+  internal struct KeyValuePair: Sendable, Hashable {
+    internal let key: String
+    internal let value: Value
+
+    /// Constructor for a metadata key-value pair.
+    /// - Parameters:
+    ///   - key: The key for the key-value pair.
+    ///   - value: The value to be associated to the given key. If it's a binary value, then the associated
+    ///   key must end in `-bin`, otherwise, this method will produce an assertion failure.
+    init(key: String, value: Value) {
+      if case .binary = value {
+        assert(key.hasSuffix("-bin"), "Keys for binary values must end in -bin")
+      }
+      self.key = key
+      self.value = value
+    }
+  }
+
+  public typealias KeyValue = (key: String, value: Value)
+
+  private var elements: [KeyValuePair]
+
+  /// The Metadata collection's capacity.
+  public var capacity: Int {
+    self.elements.capacity
+  }
+
+  /// Initialize an empty Metadata collection.
+  public init() {
+    self.elements = []
+  }
+
+  /// Reserve the specified minimum capacity in the collection.
+  /// - Parameter minimumCapacity: The minimum capacity to reserve in the collection.
+  public mutating func reserveCapacity(_ minimumCapacity: Int) {
+    self.elements.reserveCapacity(minimumCapacity)
+  }
+
+  /// Add a new key-value pair, where the value is a string.
+  /// - Parameters:
+  ///   - stringValue: The string value to be associated with the given key.
+  ///   - key: The key to be associated with the given value.
+  public mutating func addString(_ stringValue: String, forKey key: String) {
+    self.addValue(.string(stringValue), forKey: key)
+  }
+
+  /// Add a new key-value pair, where the value is binary data, in the form of `Data`.
+  /// - Parameters:
+  ///   - binaryValue: The `Data` to be associated with the given key.
+  ///   - key: The key to be associated with the given value. Must end in `-bin`.
+  public mutating func addBinary(_ binaryValue: [UInt8], forKey key: String) {
+    self.addValue(.binary(binaryValue), forKey: key)
+  }
+
+  /// Add a new key-value pair.
+  /// - Parameters:
+  ///   - value: The ``Value`` to be associated with the given key.
+  ///   - key: The key to be associated with the given value. If value is binary, it must end in `-bin`.
+  public mutating func addValue(_ value: Value, forKey key: String) {
+    self.elements.append(.init(key: key, value: value))
+  }
+
+  /// Removes all values associated with the given key.
+  /// - Parameter key: The key for which all values should be removed.
+  public mutating func removeAllValues(forKey key: String) {
+    elements.removeAll { metadataKeyValue in
+      metadataKeyValue.key == key
+    }
+  }
+
+  /// Adds a key-value pair to the collection, where the value is a string.
+  /// If there are pairs already associated to the given key, they will all be removed first, and the new pair
+  /// will be added. If no pairs are present with the given key, a new one will be added.
```

Comment:
Need to be careful with spacing here as it's semantically important for tools to render docs correctly (e.g. DocC, Xcode).

The first line should be a summary, any further discussion should be separated from the summary by a blank line.

The best guide on this is probably the API design guidelines (https://www.swift.org/documentation/api-design-guidelines/). Take a look through the details of the "Write a documentation comment" section.

This needs fixing up in a couple of other places too.

Reply:
Could you also add: `- Complexity: O(count)`? (and to the other funcs this applies too). It's generally considered good practice to document the complexity of anything that isn't O(1).

I think OpenAPI have raised the bar on docs, we should try to raise ourselves to their level so I'm making a concerted effort to do this for v2.

Reply:
How detailed do we want to be with regards to the complexity docs? Particularly, you think it's worth mentioning worst case scenario for the iterators' `next` methods? For example, in the binary values iterator, we compare keys and try to decode strings, moving on to the next key if the decoding fails. This means that worst case scenario, we'd see a complexity of `O(n*cost of trying to decode a string value)`, although it's likely this won't be as bad in the real world.

Reply:
I think keep it relatively simple. IMO the intention is so that users know not to e.g. store the result of something if it's linear rather than recomputing it each time.  For the binary iterator I'd ignore the cost of decoding but mention in the docs that values may be decoded on-demand so user can get an idea that values may be decoded.